### PR TITLE
Refactor api to make it more testeable - fixes #62

### LINKED ISSUE #62: Refactor API to make it more testeable.
**Feature description**

Is a bit hard to write tests with the current library API.

**Is your feature request related to a problem? Please describe.**
Just write tests as requested:
https://github.com/skycoin/hardware-wallet-go/pull/50#pullrequestreview-213769530

**Describe the solution you'd like**
An interface based implementation in the lib.

**Describe alternatives you've considered**
An interface based implementation in the lib.
Or write complex integration tests to tests local functionalities.

**Additional context**
https://github.com/skycoin/hardware-wallet-go/pull/50#pullrequestreview-213769530

**Possible implementation**
Just create some interface to encapsulate functionality

### PR DESCRIPTION
**Pull request by stdevAlDen**
Fixes #62

Changes:
- A simple refactor to use some other interfaces and get open the posibility of use mock objects 

Does this change need to mentioned in CHANGELOG.md?
yes

Requires testing
no

### CODE REVIEW

File: src/cli/address_gen.go [hunk 1]
```diff
@@ -44,14 +44,8 @@ func addressGenCmd() gcli.Command {
...
-			var device *deviceWallet.Device
-			switch c.String("deviceType") {
-			case "USB":
-				device = deviceWallet.NewUSBDevice()
-			case "EMULATOR":
-				device = deviceWallet.NewEmulatorDevice()
-			default:
-				log.Error("device type not set")
+			device := deviceWallet.NewDevicer(c.String("deviceType"))
+			if device == nil {
```
Comment (gz-c) [inline src/cli/address_gen.go]: These `device == nil` checks mean that the `deviceType` flag is ignored if the device is already set. That is incorrect behavior

File: src/device-wallet/helper.go [hunk 2]
```diff
@@ -17,41 +15,38 @@ import (
...
-	if err != nil {
-		return false
-	}
+type DeviceDriver interface {
+	getEmulatorDevice() (net.Conn, error)
+	getUsbDevice() (usb.Device, error)
+	sendToDeviceNoAnswer(dev io.ReadWriteCloser, chunks [][64]byte) error
+	sendToDevice(dev io.ReadWriteCloser, chunks [][64]byte) (wire.Message, error)
+	binaryWrite(message io.Writer, data interface{})
+	makeTrezorMessage(data []byte, msgID messages.MessageType) [][64]byte
```
Comment (olemis) [inline src/device-wallet/helper.go]: `makeSkyWalletMessage`

File: src/device-wallet/helper.go [hunk 3]
```diff
@@ -17,41 +15,38 @@ import (
...
-func deviceConnected(deviceType DeviceType) bool {
-	dev, err := getDevice(deviceType)
-	if dev == nil {
-		return false
-	}
-	defer dev.Close()
-	if err != nil {
-		return false
-	}
+type DeviceDriver interface {
```
Comment (gz-c) [inline src/device-wallet/helper.go]: Public interface type shouldn't have private methods

Comment (olemis) [reply src/device-wallet/helper.go]: right ...

**Comment by stdevAlDen** [comment]
> needed

Sound better, let me check...

File: src/device-wallet/helper.go (hunk 3 above)
Comment (stdevAlDen) [reply src/device-wallet/helper.go]: > Public interface type shouldn't have private methods

ok, to follow this pattern I need to make they(at least one so far now) public as could be required create a stub from here https://github.com/skycoin/hardware-wallet-go/pull/50/files#diff-e74c5dbf5b6565d473a1d8b1efd4df53R311

File: src/device-wallet/messages.go [hunk 4]
```diff
@@ -285,6 +324,87 @@ func MessagePinMatrixAck(p string) ([][64]byte, error) {
...
+func (pb *Protobuf) DecodeResponseSkycoinSignMessage(msg wire.Message) (string, error) {
+	if msg.Kind == uint16(messages.MessageType_MessageType_ResponseSkycoinSignMessage) {
+		responseSkycoinSignMessage := &messages.ResponseSkycoinSignMessage{}
+		err := proto.Unmarshal(msg.Data, responseSkycoinSignMessage)
+		if err != nil {
+			return "", err
+		}
+		return responseSkycoinSignMessage.GetSignedMessage(), nil
+	}
+	return "", fmt.Errorf("calling DecodeResponseeSkycoinSignMessage with wrong message type: %s", messages.MessageType(msg.Kind))
```
Comment (therealssj) [inline src/device-wallet/messages.go]: This interface is not needed. There is nothing that needs to be mocked

File: src/device-wallet/helper.go [hunk 5]
```diff
@@ -17,41 +15,38 @@ import (
...
+	getEmulatorDevice() (net.Conn, error)
+	getUsbDevice() (usb.Device, error)
```
Comment (therealssj) [inline src/device-wallet/helper.go]: remove these as`getDevice` should act as the gateway to get the device.

File: src/device-wallet/helper.go (hunk 2 above)
Comment (therealssj) [inline src/device-wallet/helper.go]: only these are needed in this interface.
```
SendToDeviceNoAnswer
SendToDevice
```
rest all are not functions to be mocked

File: src/device-wallet/messages.go (hunk 4 above)
Comment (stdevAlDen) [reply src/device-wallet/messages.go]: @therealssj maybe related to a different issue but, could be great let open the possibility to replace protobuf, consider this for a different issue (low priority for now).

**Comment by stdevAlDen** [comment]
> This is how I think it should be
> 
> ```
> type Device struct {
> 	Driver
> }
> 
> type DeviceDriver interface {
> 	SendToDeviceNoAnswer
> 	SendToDevice
> }
> 
> type Driver struct {
> 	DeviceType
> }
> ```

@therealssj  great, thank you very much.

**Comment by stdevAlDen** [comment]
ready for final review -> merge
thank you for your supprt guys @olemis @therealssj @gz-c , let me know if I miss any thing, the only detail I no like to much is that we have a [specific](https://github.com/simelo/skycoin-hardware-wallet-go/blob/stdevAlDen_t62_refactor_API_to_make_it_more_testeable/src/device-wallet/device-wallet.go#L44) implementation [instead of a generic interface](https://github.com/simelo/skycoin-hardware-wallet-go/blob/stdevAlDen_t62_refactor_API_to_make_it_more_testeable/src/device-wallet/helper.go#L17) here.

File: src/device-wallet/helper.go [hunk 6]
```diff
@@ -4,47 +4,58 @@ import (
...
+	// DeviceTypeEmulator use emulator
+	DeviceTypeEmulator DeviceType = 1
+	// DeviceTypeEmulatorStr string to represent DeviceTypeEmulator
+	DeviceTypeEmulatorStr string = "EMULATOR"
+	// DeviceTypeUsb use usb
+	DeviceTypeUSB DeviceType = 2
+	// DeviceTypeUSBStr string to represent DeviceTypeUSB
+	DeviceTypeUSBStr string = "USB"
+	// DeviceTypeInvalid
+	DeviceTypeInvalid DeviceType = 3
```
Comment (therealssj) [inline src/device-wallet/helper.go]: implement a string method on `DeviceType`  which returns the string representation
And use `iota` for 1,2,3  (start with iota+1)

File: src/device-wallet/device_interface_test.go [hunk 7]
```diff
@@ -12,17 +12,23 @@ import (
 	messages "github.com/skycoin/hardware-wallet-go/src/device-wallet/messages/go"
 )
 
+func testHelperGetDeviceWithBestEffort(testName string, t *testing.T) *Device {
```
Comment (therealssj) [inline src/device-wallet/device_interface_test.go]: `testHelperGetDeviceWithBestEffort(t *testing.T, testName string)`

File: src/device-wallet/messages.go (hunk 4 above)
Comment (therealssj) [reply src/device-wallet/messages.go]: move these back to `helper` package these are not `Messages`

**Comment by therealssj** [comment]
> ready for final review -> merge
> thank you for your supprt guys @olemis @therealssj @gz-c , let me know if I miss any thing, the only detail I no like to much is that we have a [specific](https://github.com/simelo/skycoin-hardware-wallet-go/blob/stdevAlDen_t62_refactor_API_to_make_it_more_testeable/src/device-wallet/device-wallet.go#L44) implementation [instead of a generic interface](https://github.com/simelo/skycoin-hardware-wallet-go/blob/stdevAlDen_t62_refactor_API_to_make_it_more_testeable/src/device-wallet/helper.go#L17) here.

Its fine to not have generics in go. Repeat code if it makes it more idiomatic

**Comment by stdevAlDen** [comment]
@therealssj done

**Comment by olemis** [comment]
@stdevAlDen if you need further changes please implement them in #44 for #38 .

**Comment by olemis** [comment]
@olemis it's #50 for #46

**Pull request merged**